source/core/constraint/unique.d

```diff
@@ -13,45 +13,177 @@ import core.sudoku.cell;
  */
 public class UniqueConstraint : Constraint
 {
-	private this(Cell[] cells...)
+	private this(Cell cell)
 	{
 		super (ConstraintType.Unique);
-		this.cells ~= cells;
+		this.cells ~= cell;
 	}
 
 
-	public static auto createSingle(Cell cell1, Cell cell2)
+	/** Connects two Cells
+	 *
+	 * Ensures `cell1` is connected to `cell2` and vice-versa
+	 * This means when a digit is checked for it's availability in `cell1`
+	 *     the validation is dependent of the digit in `cell2` and vice-versa
+	 * A Cell cannot connect to itself
+	 *
+	 * Params:
+	 *     cell1 = cell to connect
+	 *     cell2 = cell to connect
+	 *
+	 * Examples:
+	 * --------------------
+	 * Cell cell1 = new Cell(0);
+	 * Cell cell2 = new Cell(0);
+	 * UniqueConstraint.createSingle(cell1,cell2);
+	 * assert(cell1.get!UniqueConstraint.cells.equal([cell2]));
+	 * assert(cell2.get!UniqueConstraint.cells.equal([cell1]));
+	 * --------------------
+	 */
+	public static void createSingle(Cell cell1, Cell cell2)
 	{
-		if (cell1 is null || cell2 is null) return;
-		UniqueConstraint uc1 = new UniqueConstraint(cell1, cell2);
+		if (cell1 is null || cell2 is null)
```

Comment:
Why not employ the trick you used in some remove methods here?

```d
auto validCells = cells.filter!(cell => cell !is null).array;
```

Reply:
The way filter works is that it checks for a condition in a Range, and then it returns a new Range with the values for which the condition returns true. `cell1` and `cell2` aren't ranges, and to use it with this they need to be. You could do something like `[cell1,cell2].filter!(cell => cell !is null)`, but then you would also need to check whether the returned Range was empty or not which is more inefficient.
The function above filters the null objects if they exist. This is intended in case some of the Cells are null for some reason. If you are using the Grid fucntions to pass the Cells there shouldn't be a problem with this, however if for some reason you need to pass custom Cells, then it could be a problem, ence this filter. It just means, all the valid Cells are still connected. `UniqueConstraint.createInterconnected([new Cell(1), new Cell(0), null, new Cell(0)]); // no error`